encode.go

```diff
@@ -75,11 +76,28 @@ func emitCopy(dst []byte, offset, length int) int {
 	return i
 }
 
+var encPool = sync.Pool{New: func() interface{} { return new(encoder) }}
+
 // Encode returns the encoded form of src. The returned slice may be a sub-
 // slice of dst if dst was large enough to hold the entire encoded block.
 // Otherwise, a newly allocated slice will be returned.
 // It is valid to pass a nil dst.
 func Encode(dst, src []byte) []byte {
+	var e = encPool.Get().(*encoder)
+	dst = e.Encode(dst, src)
+	encPool.Put(e)
+	return dst
+}
+
+const tableBits = 14             // Bits used in the table
+const tableSize = 1 << tableBits // Size of the table
+
+type encoder struct {
+	table [tableSize]int32
+	cur   int
```

Comment:
I find it surprising that it's a pool of array-and-int instead of just an array.